src/compas_fab/backends/pybullet/client.py

```diff
@@ -0,0 +1,410 @@
+from .utils import LOG
+from .utils import redirect_stdout
+from .conversions import frame_from_pose
+from .conversions import pose_from_frame
+
+pybullet = LazyLoader('pybullet', globals(), 'pybullet')
+
+
+__all__ = [
+    'PyBulletClient',
+]
+
+
+class PyBulletBase(object):
+    def __init__(self, use_gui):
+        self.client_id = None
+        self.use_gui = use_gui
+
+    def connect(self, shadows=True, color=None, width=None, height=None):
+        # Shared Memory: execute the physics simulation and rendering in a separate process
+        # https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/examples/vrminitaur.py#L7
+        self.detect_display()
+        method = pybullet.GUI if self.use_gui else pybullet.DIRECT
+        options = self.compose_options(color, width, height)
+        with redirect_stdout():
+            self.client_id = pybullet.connect(method, options=options)
+        if 0 > self.client_id:
+            raise Exception('Error in establishing connection with PyBullet.')
+        if self.use_gui:
+            self.configure_debug_visualizer(shadows)
+
+    def detect_display(self):
+        if self.use_gui and system != 'darwin' and not compas.is_windows() and ('DISPLAY' not in os.environ):
+            self.use_gui = False
+            print('No display detected! Continuing without GUI.')
+
+    @staticmethod
+    def compose_options(color, width, height):
+        options = ''
+        if color is not None:
+            options += '--background_color_red={} --background_color_green={} --background_color_blue={}'.format(*color)
+        if width is not None:
+            options += '--width={}'.format(width)
+        if height is not None:
+            options += '--height={}'.format(height)
+        return options
+
+    def configure_debug_visualizer(self, shadows):
+        pybullet.configureDebugVisualizer(pybullet.COV_ENABLE_GUI, False, physicsClientId=self.client_id)
+        pybullet.configureDebugVisualizer(pybullet.COV_ENABLE_TINY_RENDERER, False, physicsClientId=self.client_id)
+        pybullet.configureDebugVisualizer(pybullet.COV_ENABLE_RGB_BUFFER_PREVIEW, False, physicsClientId=self.client_id)
+        pybullet.configureDebugVisualizer(pybullet.COV_ENABLE_DEPTH_BUFFER_PREVIEW, False, physicsClientId=self.client_id)
+        pybullet.configureDebugVisualizer(pybullet.COV_ENABLE_SEGMENTATION_MARK_PREVIEW, False, physicsClientId=self.client_id)
+        pybullet.configureDebugVisualizer(pybullet.COV_ENABLE_SHADOWS, shadows, physicsClientId=self.client_id)
+
+    def disconnect(self):
+        with redirect_stdout():
+            return pybullet.disconnect(physicsClientId=self.client_id)
+
+    @property
+    def is_connected(self):
+        """Indicates whether the client has an active connection.
+
+        Returns
+        -------
+        :obj:`bool`
+            ``True`` if connected, ``False`` otherwise.
+        """
+        return self.check_connection() == 1
```

Comment:
This fails if it's never connected, because `client_id` is `None`:
```
>>> client = PyBulletClient()
>>> client.is_connected
pybullet build time: Jul  8 2020 13:56:47
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "c:\users\gcasas\eth\projects\research\compas_fab\src\compas_fab\backends\pybullet\client.py", line 95, in is_connected
    return self.check_connection() == 1
  File "c:\users\gcasas\eth\projects\research\compas_fab\src\compas_fab\backends\pybullet\client.py", line 98, in check_connection
    return pybullet.getConnectionInfo(physicsClientId=self.client_id)['isConnected']
TypeError: an integer is required (got type NoneType)
```

Reply:
In testing my fix for this, I noticed the following:  
<img width="550" alt="Screenshot 2020-07-09 at 15 14 57" src="https://user-images.githubusercontent.com/57254617/87044627-19134e80-c1f7-11ea-88e1-96b3db2af949.png">
Should I try to fix this? I fear it might be tricky...

Reply:
I'm not sure I understand what you are pointing out on your comment. The output looks alright.. doesn't it? (except the fact that the printouts are all after the `.disconnect()` instead of after each line but I assume that is totally unrelated)

Reply:
Yeah, it's that the output comes after `disconnect`.  I'm sure it has something to do with `redirect_stdout` and is unrelated to `is_connected`, but this is just how I noticed it.

Reply:
ah yes, it is probably related to buffering flags of `stdout` (there's a comment about it in the `finally` block of `redirect_stdout`). Maybe you can try to disable buffering in line 36 with something like `sys.stdout = os.fdopen(fd, 'w', buffering=0)`.

Reply:
`ValueError: can't have unbuffered text I/O`

Reply:
😬